File: Activity_selection_problem/activity_selection_problem.cpp
/*

Author: Aditya Upadhye

*/ 


#include <iostream>
#include <map>
#include <vector>


bool check_start_time(const int& start_time)
{

	if (start_time < 0)
		return false;
	else 
		return true;
}


bool check_finish_time(const int& start_time, const int& finish_time)
{

	if (finish_time <= start_time)
		return false;
	else 
		return true;

}


void activity_selector(const std::multimap<int, int>& act_list,const std::multimap<int, int>& act_count)
{


	std::vector<int> selected_activities;

	int prev_ft = act_list.begin() -> second;

	auto itr_act_count = act_count.begin();

	for (auto key: act_list)
	{

		if (key.second >= prev_ft)
		{

			selected_activities.push_back(itr_act_count -> second);
			prev_ft = key.first;
		}


		itr_act_count++; 
        }

	if (selected_activities.size())
		std::cout << "The following activities will be selected: " << std::endl;

	for (auto& no : selected_activities)
	{
		std::cout << no << " ";

	}
	std::cout << std::endl;

}


int main()
{

	std::multimap<int, int> activity_list;
	std::multimap<int, int> activity_no;

	int act_count = 0;

	std::cout << "Enter the no of activities: " << std::endl;
	std::cin >> act_count;
	int act_no = 1;

	while (act_count)
	{

		int start_time = -1;
		int finish_time = -1;

		std::cout << "Enter the start time: " << std::endl;
		std::cin >> start_time;

		while (!check_start_time(start_time))
		{

			std::cout << "Enter a valid start time: " << std::endl;
			std::cin >> start_time;

		}


		while (!check_finish_time(start_time, finish_time))
		{

			std::cout << "Enter a valid finish time: " << std::endl;
			std::cin >> finish_time;

		}

		activity_list.insert({finish_time, start_time});
		activity_no.insert({finish_time, act_no});
		act_count--;
		++act_no;

	}

	activity_selector(activity_list, activity_no);

	return 0;

}

File: Activity_selection_problem/activity_selection_problem.py
def check_start_time(start_time):
    
    if start_time >= 0:
        return True
    else:
        return False    


def check_finish_time(st, ft):
    
    if ft >= st:
        return True
    else:
        return False


def activity_selector(activity, activity_count):

    selected_activities = []
    
    for key, value in activity_count.items():
    
        if selected_activities:
            
            if activity[key][0] >= activity[selected_activities[-1]][1]:
                
                selected_activities.append(key)
        else:
            
            selected_activities.append(key)
    
    if selected_activities:
        print('\n\nThe following activities will be selected: ') 
    
    for act in selected_activities:
        print(act)
    
###################

no_of_act = int(input('Enter the number of activities: '))

activity = {}
activity_count = {}
activity_no = 1

while (no_of_act):
    
    start_time = int(input('Enter a valid start time: '))
    
    while not check_start_time(start_time):
        
        start_time = int(input('Re-enter a valid start time: '))
        
    finish_time = int(input('Enter a valid finish time: '))    
    
    while not check_finish_time(start_time, finish_time):
    
        finish_time = int(input('Re-enter a valid finish time: ')) 
    
    activity[activity_no] = (start_time, finish_time)
    activity_count[activity_no] = finish_time
    
    activity_no += 1
    
    no_of_act -= 1

activity_count = dict(sorted(activity_count.items(), key=lambda value: value[1]))

activity_selector(activity, activity_count)

       
